File: modules/math/src/main/java/com/flowpowered/math/vector/Vec3i.java
package com.flowpowered.math.vector;

import java.io.Serializable;
import java.lang.Override;

import com.flowpowered.math.AllocationTracker;
import com.flowpowered.math.GenericMath;
import com.flowpowered.math.HashFunctions;
import org.jetbrains.annotations.NotNull;

public class Vec3i implements Vectori, Vec3, Comparable<Vec3i>, Serializable, Cloneable {
    private static final long serialVersionUID = 1;
    private static final FlowPool<Vec3i> pool = new FlowPool<Vec3i>(3, FlowPool.BITS_3D) {
        @NotNull
        protected Vec3i hit(Vec3i value) {
            AllocationTracker.cacheHitsVec3i++;
            return value;
        }

        @NotNull
        protected Vec3i create(int x, int y, int z, int w) {
            return new Vec3i(x, y, z);
        }
    };

    public static final Vec3i ZERO = new Vec3i(0, 0, 0);
    public static final Vec3i UNIT_X = new Vec3i(1, 0, 0);
    public static final Vec3i UNIT_Y = new Vec3i(0, 1, 0);
    public static final Vec3i UNIT_Z = new Vec3i(0, 0, 1);
    public static final Vec3i ONE = new Vec3i(1, 1, 1);
    public static final Vec3i RIGHT = UNIT_X;
    public static final Vec3i UP = UNIT_Y;
    public static final Vec3i FORWARD = UNIT_Z;
    private final int x;
    private final int y;
    private final int z;
    private transient volatile boolean hashed = false;
    private transient volatile int hashCode = 0;

    public Vec3i() {
        this(0, 0, 0);
    }

    public Vec3i(@NotNull Vec2i v) {
        this(v, 0);
    }

    public Vec3i(@NotNull Vec2i v, double z) {
        this(v, GenericMath.floor(z));
    }

    public Vec3i(@NotNull Vec2i v, int z) {
        this(v.getX(), v.getY(), z);
    }

    public Vec3i(@NotNull Vec3i v) {
        this(v.x, v.y, v.z);
    }

    public Vec3i(@NotNull Vec4i v) {
        this(v.getX(), v.getY(), v.getZ());
    }

    public Vec3i(@NotNull VecNi v) {
        this(v.get(0), v.get(1), v.size() > 2 ? v.get(2) : 0);
    }

    public Vec3i(double x, double y, double z) {
        this(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    public Vec3i(int x, int y, int z) {
        this.x = x;
        this.y = y;
        this.z = z;
        AllocationTracker.instancesVec3i++;
    }

    public int getX() {
        return x;
    }
    /** Operator function for Kotlin */
    public int component1() {
        return getX();
    }

    public int getY() {
        return y;
    }
    /** Operator function for Kotlin */
    public int component2() {
        return getY();
    }

    public int getZ() {
        return z;
    }
    /** Operator function for Kotlin */
    public int component3() {
        return getZ();
    }

    @NotNull
    public Vec3i add(@NotNull Vec3i v) {
        return add(v.x, v.y, v.z);
    }
    /** Operator function for Kotlin */
    @NotNull
    public Vec3i plus(@NotNull Vec3i v) {
        return add(v);
    }

    @NotNull
    public Vec3i add(double x, double y, double z) {
        return add(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i add(int x, int y, int z) {
        return new Vec3i(this.x + x, this.y + y, this.z + z);
    }

    @NotNull
    public Vec3i sub(@NotNull Vec3i v) {
        return sub(v.x, v.y, v.z);
    }
    /** Operator function for Kotlin */
    @NotNull
    public Vec3i minus(@NotNull Vec3i v) {
        return sub(v);
    }

    @NotNull
    public Vec3i sub(double x, double y, double z) {
        return sub(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i sub(int x, int y, int z) {
        return new Vec3i(this.x - x, this.y - y, this.z - z);
    }

    @NotNull
    public Vec3i mul(double a) {
        return mul(GenericMath.floor(a));
    }
    /** Operator function for Kotlin */
    @NotNull
    public Vec3i times(double a) {
        return mul(a);
    }

    @NotNull
    @Override
    public Vec3i mul(int a) {
        return mul(a, a, a);
    }
    /** Operator function for Kotlin */
    @NotNull
    @Override
    public Vec3i times(int a) {
        return mul(a);
    }

    @NotNull
    public Vec3i mul(@NotNull Vec3i v) {
        return mul(v.x, v.y, v.z);
    }
    /** Operator function for Kotlin */
    @NotNull
    public Vec3i times(@NotNull Vec3i v) {
        return mul(v);
    }

    @NotNull
    public Vec3i mul(double x, double y, double z) {
        return mul(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i mul(int x, int y, int z) {
        return new Vec3i(this.x * x, this.y * y, this.z * z);
    }

    @NotNull
    public Vec3i div(double a) {
        return div(GenericMath.floor(a));
    }

    @NotNull
    @Override
    public Vec3i div(int a) {
        return div(a, a, a);
    }

    @NotNull
    public Vec3i div(@NotNull Vec3i v) {
        return div(v.x, v.y, v.z);
    }

    @NotNull
    public Vec3i div(double x, double y, double z) {
        return div(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i div(int x, int y, int z) {
        return new Vec3i(this.x / x, this.y / y, this.z / z);
    }

    public int dot(@NotNull Vec3i v) {
        return dot(v.x, v.y, v.z);
    }

    public int dot(double x, double y, double z) {
        return dot(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    public int dot(int x, int y, int z) {
        return this.x * x + this.y * y + this.z * z;
    }

    @NotNull
    public Vec3i project(@NotNull Vec3i v) {
        return project(v.x, v.y, v.z);
    }

    @NotNull
    public Vec3i project(double x, double y, double z) {
        return project(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i project(int x, int y, int z) {
        final int lengthSquared = x * x + y * y + z * z;
        if (lengthSquared == 0) {
            throw new ArithmeticException("Cannot project onto the zero vector");
        }
        final float a = (float) dot(x, y, z) / lengthSquared;
        return new Vec3i(a * x, a * y, a * z);
    }

    @NotNull
    public Vec3i cross(@NotNull Vec3i v) {
        return cross(v.x, v.y, v.z);
    }

    @NotNull
    public Vec3i cross(double x, double y, double z) {
        return cross(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i cross(int x, int y, int z) {
        return new Vec3i(this.y * z - this.z * y, this.z * x - this.x * z, this.x * y - this.y * x);
    }

    @NotNull
    public Vec3i pow(double pow) {
        return pow(GenericMath.floor(pow));
    }

    @NotNull
    @Override
    public Vec3i pow(int power) {
        return new Vec3i(Math.pow(x, power), Math.pow(y, power), Math.pow(z, power));
    }

    @NotNull
    @Override
    public Vec3i abs() {
        return new Vec3i(Math.abs(x), Math.abs(y), Math.abs(z));
    }

    @NotNull
    @Override
    public Vec3i negate() {
        return new Vec3i(-x, -y, -z);
    }
    /** Operator function for Kotlin */
    @NotNull
    @Override
    public Vec3i unaryMinus() {
        return negate();
    }

    @NotNull
    public Vec3i min(@NotNull Vec3i v) {
        return min(v.x, v.y, v.z);
    }

    @NotNull
    public Vec3i min(double x, double y, double z) {
        return min(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i min(int x, int y, int z) {
        return new Vec3i(Math.min(this.x, x), Math.min(this.y, y), Math.min(this.z, z));
    }

    @NotNull
    public Vec3i max(@NotNull Vec3i v) {
        return max(v.x, v.y, v.z);
    }

    @NotNull
    public Vec3i max(double x, double y, double z) {
        return max(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    @NotNull
    public Vec3i max(int x, int y, int z) {
        return new Vec3i(Math.max(this.x, x), Math.max(this.y, y), Math.max(this.z, z));
    }

    public int distanceSquared(@NotNull Vec3i v) {
        return distanceSquared(v.x, v.y, v.z);
    }

    public int distanceSquared(double x, double y, double z) {
        return distanceSquared(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    public int distanceSquared(int x, int y, int z) {
        final int dx = this.x - x;
        final int dy = this.y - y;
        final int dz = this.z - z;
        return dx * dx + dy * dy + dz * dz;
    }

    public float distance(@NotNull Vec3i v) {
        return distance(v.x, v.y, v.z);
    }

    public float distance(double x, double y, double z) {
        return distance(GenericMath.floor(x), GenericMath.floor(y), GenericMath.floor(z));
    }

    public float distance(int x, int y, int z) {
        return (float) Math.sqrt(distanceSquared(x, y, z));
    }

    @Override
    public int lengthSquared() {
        return x * x + y * y + z * z;
    }

    @Override
    public float length() {
        return (float) Math.sqrt(lengthSquared());
    }

    /**
     * Returns the axis with the minimal value.
     *
     * @return {@link int} axis with minimal value
     */
    @Override
    public int getMinAxis() {
        return x < y ? (x < z ? 0 : 2) : (y < z ? 1 : 2);
    }

    /**
     * Returns the axis with the maximum value.
     *
     * @return {@link int} axis with maximum value
     */
    @Override
    public int getMaxAxis() {
        return x < y ? (y < z ? 2 : 1) : (x < z ? 2 : 0);
    }

    @NotNull
    public Vec2i toVec2() {
        return new Vec2i(this);
    }

    @NotNull
    public Vec2i toVec2(boolean useZ) {
        return new Vec2i(x, useZ ? z : y);
    }

    @NotNull
    public Vec4i toVec4() {
        return toVec4(0);
    }

    @NotNull
    public Vec4i toVec4(double w) {
        return toVec4(GenericMath.floor(w));
    }

    @NotNull
    public Vec4i toVec4(int w) {
        return new Vec4i(this, w);
    }

    @NotNull
    public VecNi toVecN() {
        return new VecNi(this);
    }

    @NotNull
    @Override
    public int[] toArray() {
        return new int[]{x, y, z};
    }

    @NotNull
    @Override
    public Vec3d toDouble() {
        return new Vec3d(x, y, z);
    }

    @Override
    public double getXd() {
        return (double) getX();
    }

    @Override
    public double getYd() {
        return (double) getY();
    }

    @Override
    public double getZd() {
        return (double) getZ();
    }

    @NotNull
    @Override
    public net.minecraft.util.math.Vec3d toMC() {
        return new net.minecraft.util.math.Vec3d(getXd(), getYd(), getZd());
    }

    @NotNull
    @Override
    public Vec3f toFloat() {
        return new Vec3f(x, y, z);
    }

    @Override
    public float getXf() {
        return (float) getX();
    }

    @Override
    public float getYf() {
        return (float) getY();
    }

    @Override
    public float getZf() {
        return (float) getZ();
    }

    @NotNull
    @Override
    public Vec3l toLong() {
        return new Vec3l(x, y, z);
    }

    @Override
    public long getXl() {
        return (long) getX();
    }

    @Override
    public long getYl() {
        return (long) getY();
    }

    @Override
    public long getZl() {
        return (long) getZ();
    }

    @NotNull
    @Override
    public Vec3i toInt() {
        return this;
    }

    @Override
    public int getXi() {
        return (int) getX();
    }

    @Override
    public int getYi() {
        return (int) getY();
    }

    @Override
    public int getZi() {
        return (int) getZ();
    }

    @NotNull
    @Override
    public Number getXn() {
        return getX();
    }

    @NotNull
    @Override
    public Number getYn() {
        return getY();
    }

    @NotNull
    @Override
    public Number getZn() {
        return getZ();
    }

    @Override
    public int compareTo(@NotNull Vec3i v) {
        return (int) (lengthSquared() - v.lengthSquared());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Vec3i)) {
            return false;
        }
        final Vec3i Vec3 = (Vec3i) o;
        if (Vec3.x != x) {
            return false;
        }
        if (Vec3.y != y) {
            return false;
        }
        if (Vec3.z != z) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        if (!hashed) {
            hashCode = ((HashFunctions.hash(x) * 211 + HashFunctions.hash(y)) * 97 + HashFunctions.hash(z));
            hashed = true;
        }
        return hashCode;
    }

    @NotNull
    @Override
    public Vec3i clone() {
        return new Vec3i(this);
    }

    @NotNull
    @Override
    public String toString() {
        return "(" + x + ", " + y + ", " + z + ")";
    }

    @NotNull
    public static Vec3i from(int n) {
         return n == 0 ? ZERO : new Vec3i(n, n, n);
    }

    @NotNull
    public static Vec3i from(int x, int y, int z) {
         return x == 0 && y == 0 && z == 0 ? ZERO : new Vec3i(x, y, z);
    }

    public static Vec3i createPooled(int x, int y, int z) {
        return pool.getPooled(x, y, z, 0);
    }
}

File: modules/math/src/main/kotlin/com/teamwizardry/librarianlib/math/vector/Vec3i.kt
package com.teamwizardry.librarianlib.math.vector

import net.minecraft.util.math.BlockPos
import kotlin.math.absoluteValue
import net.minecraft.util.math.Vec3d as MCVec3d

interface Vec3i: Vec3 {
    val x: Int
    val y: Int
    val z: Int

    @JvmDefault override val xd: Double get() = x.toDouble()
    @JvmDefault override val yd: Double get() = y.toDouble()
    @JvmDefault override val zd: Double get() = z.toDouble()
    @JvmDefault override val xf: Float get() = x.toFloat()
    @JvmDefault override val yf: Float get() = y.toFloat()
    @JvmDefault override val zf: Float get() = z.toFloat()
    @JvmDefault override val xi: Int get() = x
    @JvmDefault override val yi: Int get() = y
    @JvmDefault override val zi: Int get() = z

    @JvmDefault override fun add(x: Int, y: Int, z: Int): Vec3i = getPooled(this.x + x, this.y + y, this.z + z)
    @JvmDefault override fun add(x: Float, y: Float, z: Float): Vec3d = Vec3d.getPooled(this.x + x, this.y + y, this.z + z)
    @JvmDefault override fun add(x: Double, y: Double, z: Double): Vec3d = Vec3d.getPooled(this.x + x, this.y + y, this.z + z)

    @JvmDefault override fun sub(x: Int, y: Int, z: Int): Vec3i = getPooled(this.x - x, this.y - y, this.z - z)
    @JvmDefault override fun sub(x: Float, y: Float, z: Float): Vec3d = Vec3d.getPooled(this.x - x, this.y - y, this.z - z)
    @JvmDefault override fun sub(x: Double, y: Double, z: Double): Vec3d = Vec3d.getPooled(this.x - x, this.y - y, this.z - z)

    @JvmDefault override fun mul(a: Int): Vec3i = getPooled(this.x * a, this.y * a, this.z * a)
    @JvmDefault override fun mul(a: Float): Vec3d = Vec3d.getPooled(this.x * a, this.y * a, this.z * a)
    @JvmDefault override fun mul(a: Double): Vec3d = Vec3d.getPooled(this.x * a, this.y * a, this.z * a)
    @JvmDefault override fun mul(x: Int, y: Int, z: Int): Vec3i = getPooled(this.x * x, this.y * y, this.z * z)
    @JvmDefault override fun mul(x: Float, y: Float, z: Float): Vec3d = Vec3d.getPooled(this.x * x, this.y * y, this.z * z)
    @JvmDefault override fun mul(x: Double, y: Double, z: Double): Vec3d = Vec3d.getPooled(this.x * x, this.y * y, this.z * z)

    @JvmDefault override fun div(a: Int): Vec3d = Vec3d.getPooled(this.xd / a, this.yd / a, this.zd / a)
    @JvmDefault override fun div(a: Float): Vec3d = Vec3d.getPooled(this.x / a, this.y / a, this.z / a)
    @JvmDefault override fun div(a: Double): Vec3d = Vec3d.getPooled(this.x / a, this.y / a, this.z / a)
    @JvmDefault override fun div(x: Int, y: Int, z: Int): Vec3d = Vec3d.getPooled(this.xd / x, this.yd / y, this.zd / z)
    @JvmDefault override fun div(x: Float, y: Float, z: Float): Vec3d = Vec3d.getPooled(this.x / x, this.y / y, this.z / z)
    @JvmDefault override fun div(x: Double, y: Double, z: Double): Vec3d = Vec3d.getPooled(this.x / x, this.y / y, this.z / z)

    @JvmDefault override fun pow(a: Int): Vec3i = getPooled(
        Math.pow(x.toDouble(), a.toDouble()).toInt(),
        Math.pow(y.toDouble(), a.toDouble()).toInt(),
        Math.pow(z.toDouble(), a.toDouble()).toInt()
    )
    @JvmDefault override fun pow(a: Float): Vec3d = Vec3d.getPooled(Math.pow(xd, a.toDouble()), Math.pow(yd, a.toDouble()), Math.pow(zd, a.toDouble()))
    @JvmDefault override fun pow(a: Double): Vec3d = Vec3d.getPooled(Math.pow(xd, a), Math.pow(yd, a), Math.pow(zd, a))

    @JvmDefault override fun ceil(): Vec3i = this
    @JvmDefault override fun floor(): Vec3i = this
    @JvmDefault override fun round(): Vec3i = this
    @JvmDefault override fun abs(): Vec3i = getPooled(x.absoluteValue, y.absoluteValue, z.absoluteValue)
    @JvmDefault override fun negate(): Vec3i = getPooled(-x, -y, -z)

    @JvmDefault override fun min(x: Int, y: Int, z: Int): Vec3i = getPooled(
        Math.min(this.xd, x.toDouble()).toInt(),
        Math.min(this.yd, y.toDouble()).toInt(),
        Math.min(this.zd, z.toDouble()).toInt()
    )
    @JvmDefault override fun min(x: Float, y: Float, z: Float): Vec3d = Vec3d.getPooled(Math.min(this.xd, x.toDouble()), Math.min(this.yd, y.toDouble()), Math.max(this.zd, z.toDouble()))
    @JvmDefault override fun min(x: Double, y: Double, z: Double): Vec3d = Vec3d.getPooled(Math.min(this.xd, x), Math.min(this.yd, y), Math.max(this.zd, z))
    @JvmDefault override fun max(x: Int, y: Int, z: Int): Vec3i = getPooled(
        Math.max(this.xd, x.toDouble()).toInt(),
        Math.max(this.yd, y.toDouble()).toInt(),
        Math.max(this.zd, z.toDouble()).toInt()
    )
    @JvmDefault override fun max(x: Float, y: Float, z: Float): Vec3d = Vec3d.getPooled(Math.max(this.xd, x.toDouble()), Math.max(this.yd, y.toDouble()), Math.max(this.zd, z.toDouble()))
    @JvmDefault override fun max(x: Double, y: Double, z: Double): Vec3d = Vec3d.getPooled(Math.max(this.xd, x), Math.max(this.yd, y), Math.max(this.zd, z))

    @JvmDefault override fun toDouble(): Vec3d = Vec3d.getPooled(xd, yd, zd)
    @JvmDefault override fun toInt(): Vec3i = this
    @JvmDefault override fun toMC(): MCVec3d = Vec3d.getPooledMC(xd, yd, zd)
    @JvmDefault override fun toBlockPos(): BlockPos = Vec3i.getPooledMC(xi, yi, zi)

    @JvmDefault fun add(v: Vec3i): Vec3i = add(v.x, v.y, v.z)
    @JvmDefault fun mul(v: Vec3i): Vec3i = mul(v.x, v.y, v.z)
    @JvmDefault fun sub(v: Vec3i): Vec3i = sub(v.x, v.y, v.z)
    @JvmDefault fun min(v: Vec3i): Vec3i = min(v.x, v.y, v.z)
    @JvmDefault fun max(v: Vec3i): Vec3i = max(v.x, v.y, v.z)
    @JvmDefault override fun cross(x: Int, y: Int, z: Int): Vec3i = getPooled(
        y * z - z * y,
        z * x - x * z,
        x * y - y * x
    )
    @JvmDefault @JvmSynthetic operator fun plus(v: Vec3i): Vec3i = add(v.xi, v.yi, v.zi)
    @JvmDefault @JvmSynthetic operator fun minus(v: Vec3i): Vec3i = sub(v.xi, v.yi, v.zi)
    @JvmDefault @JvmSynthetic operator fun times(v: Vec3i): Vec3i = mul(v.xi, v.yi, v.zi)
    @JvmDefault @JvmSynthetic override operator fun unaryMinus(): Vec3i = negate()
    @JvmDefault @JvmSynthetic override operator fun times(a: Int): Vec3i = mul(a)

    companion object {
        private val pool = object: VectorPool<Vec3i>(3, VectorPool.BITS_3D) {
            override fun create(x: Int, y: Int, z: Int, w: Int): Vec3i {
                return Impl(x, y, z)
            }

            override fun hit(value: Vec3i): Vec3i {
                return value
            }
        }
        private val mcPool = object: VectorPool<BlockPos>(3, VectorPool.BITS_3D) {
            override fun create(x: Int, y: Int, z: Int, w: Int): BlockPos {
                return BlockPos(x, y, z)
            }

            override fun hit(value: BlockPos): BlockPos {
                return value
            }
        }

        @JvmStatic
        fun getPooled(x: Int, y: Int, z: Int): Vec3i = pool.getPooled(x, y, z, 0)

        @JvmStatic
        fun getPooledMC(x: Int, y: Int, z: Int): BlockPos = mcPool.getPooled(x, y, z, 0)
    }

    private class Impl(override val x: Int, override val y: Int, override val z: Int): Vec3i
}
